docs/threat_models/smart_contracts/multipool_autoswap/mpas_sequence.puml

```diff
@@ -0,0 +1,182 @@
+@startuml
+'https://plantuml.com/sequence-diagram
+header Agoric Operating Systems
+footer Page %page% of %lastpage%
+title Multipool Auto Swap Sequence Diagram
+
+autonumber
+actor "MPAS Creator" as masc #lime
+actor "Pool Adder N" as pa #yellow
+actor "Pool User" as pu
+control "Zoe" as zoe
+database "MPAS Contract" as mpas
+
+== MPAS Creation ==
+masc -> mpas: Create instance
+return ""creatorFacet""
+== Pool Addition SML:BCK==
+autonumber 10
+pa -> mpas: ""addPool()""
+create database "Pool SML:BUCK" as pool
+mpas -> pool: Create pool for SML -> BUCK
+mpas --> pa: ""Issuer""
+== Pool Addition MOOL:BCK==
+autonumber 10
+pa -> mpas: ""addPool()""
+create database "Pool MOOL:BUCK" as pool2
+mpas -> pool2: Create pool for MOOL -> BUCK
+mpas --> pa: ""Issuer""
+
+
+== Duplicate Pool Addition MOOL:BCK==
+autonumber 30
+pa -> mpas: ""addPool()""
+mpas -->x pa: pool for MOOL:BCK already exists
+newpage
+autonumber 50
+== Swap Request ==
+group asking for the current price UNK:BUCK
+pu -> mpas: ""getOutputPrice(UNK:BUCK)""
+mpas -->x pu: ""brands were not recognized""
+end
+group asking for the current price MOOL:BUCK
+
+
+pu -> mpas: ""getOutputPrice(MOOL:BUCK)""
+mpas -> pool2: ""getPriceGivenRequiredOutput(MOOL:BUCK)""
+pool2 --> mpas: ""quote""
+mpas --> pu: ""quote""
+end
+
+group making a proposal
+pu -> pu: create proposal for SML:BUCK as ""proposal""
+
+end
+
+group requesting a SwapOut an invitation
+pu -> mpas: ""makeSwapOutInvitation()""
+return ""invitation""
+
+end
+
+group making an offer
+pu -> pu: withdraw funds from purse of SML into ""payment""
+pu -> zoe: ""offer(invitation, proposal, payment))""
+zoe -> zoe: validate offer
+create control "userSeat" as usA
+zoe -> usA: create ""userSeat""
+usA --> zoe: ""userSeat"" ref
+zoe -> pool: ""getPoolSeat()""
+create control "poolSeat" as psA
+pool -> psA: create ""poolSeat"" ref
+psA --> zoe: ""poolSeat"" ref
+zoe -> psA: add/remove funds
+zoe -> usA: add/remove funds
+zoe -> zoe: ""reallocate(seat, poolSeat)""
+
+usA --> pu: ""userSeat"" ref
```

Comment:
I would put UserSeat in a combined (vertical) box with Zoe. Each UserSeat is paired with a ZcfSeat; they're two objects with access to the same state. This diagram shows only one of the zcfSeats that interact during a swap. There's a seat (pair) for each pool, and there's a seat (named `seat` in `swap.js`) that represents the assets the user's assets. The pool seats are persistent through the life of the contract, the user seats are new for each swap (created by the invitation as you show.)


I think `box`, `endbox` lets you draw a box.  see "Participants encompass" [here ](https://plantuml.com/sequence-diagram).

Reply:
I moved the `userSeat` to be closer to Zoe to hopefully capture this.